### PR TITLE
Check for dataContext to avoid recyclingTemplate when unneeded

### PR DESCRIPTION
**Pull request by NicolasChampagne**
## PR Type
What kind of change does this PR introduce?
 Bugfix 

## What is the current behavior?
Would recycle template even when unneeded

## What is the new behavior?
Won't recycle template even when unneeded

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences
- [ ] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [x] Associated with an issue (GitHub or internal)

Internal Issue (If applicable):
https://nventive.visualstudio.com/Umbrella/_workitems/edit/144025

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs [hunk 1]
```diff
@@ -240,9 +240,12 @@ private void OnParentChanged(object instance, object key, DependencyObjectParent
 
 		internal static void PropagateOnTemplateReused(object instance)
 		{
-			if (instance is IFrameworkTemplatePoolAware a)
+			if (instance is IFrameworkTemplatePoolAware a && instance is FrameworkElement b)
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs]: Give these variables more meaningful names

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs [hunk 2]
```diff
@@ -240,9 +240,12 @@ private void OnParentChanged(object instance, object key, DependencyObjectParent
...
 			{
-				a.OnTemplateRecycled();
+				if (b.DataContext == null)
```
Comment (davidjohnoliver) [inline src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs]:
```suggestion
				// If DataContext is not null, it means it has been explicitly set (not inherited). Resetting the view could push an invalid value through 2-way binding in this case.
				if (b.DataContext == null)
```

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs (hunk 1 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs]: Use `IFrameworkElement` rather than `FrameworkElement`

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs (hunk 2 above)
Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs]: This is a case that may need to be tested by a unit test. Also, testing for null may require determining the precedence of the value.

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs]: Agreed for the test. 

@jeromelaban  Can you elaborate on why we would test for precedence in this case? At this point the template is detached from the visual tree, so if `DataContext` is non-null then we know it has been explicitly set.

**Comment by NicolasChampagne** [comment]
@davidjohnoliver any other necessary changes?

**Comment by davidjohnoliver** [comment]
> @davidjohnoliver any other necessary changes?

Jerome's comment:
> This is a case that may need to be tested by a unit test.

File: src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs [hunk 3]
```diff
@@ -240,9 +240,10 @@ private void OnParentChanged(object instance, object key, DependencyObjectParent
 
 		internal static void PropagateOnTemplateReused(object instance)
 		{
-			if (instance is IFrameworkTemplatePoolAware a)
+			// If DataContext is not null, it means it has been explicitly set (not inherited). Resetting the view could push an invalid value through 2-way binding in this case.
```
Comment (carldebilly) [inline src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs]: What happen if the DataContext is set directly on one of the children?

Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/FrameworkTemplatePool.cs]: `OnTemplateRecycled` is called individually for each view in the subtree, so it shouldn't matter.

**Comment by NicolasChampagne** [comment]
Bad merge, dropping pr
